### ISSUE TITLE
question about SLACK

### ISSUE THREAD

**Issue by eujungkim** [cloudfoundry-incubator/cf-abacus#870]
I had similar problem like below issue. (In the start of month, renewer failed to renew data.)
https://github.com/cloudfoundry-incubator/cf-abacus/issues/700

So I changed SLACK value to 1M in the renewer, accumulator, aggregator and reporting.
Then the renewer worked well.

But to make monthly reporting to large org failed sometimes with 504 gateway time-out. And abacus-usage-reporting's CPU becomes  over 100% when requested monthly reporting.
It seems like https://github.com/cloudfoundry-incubator/cf-abacus/issues/646 and I use commit version 1648da0.
I don't know whether main reason of time-out is large org or big SLACK.

Now, I wonder how much SLACK size effect to performance and is 1M proper value to abacus.
Is it ok to keep SLACK value to 1M for performance?
And should I keep SLACK value to 1M to resolve upper renewer problem? Or just increase little than default value(10m)?
And should I keep same SLACK value to the renewer, accumulator, aggregator and reporting?

**Comment by hsiliev** [comment]
@eujungkim Setting `SLACK=1M` can get Abacus running but this is definitely not ok for production. App bridge will transfer all events from 1M back, but if an app started previous month is stopped we'll get negative usage, because:
* Abacus does not have info for previous month
* CC stores only 1M of history
* we'll subtract the stopped app GBh from the org/space that has 0 GBh

The recommended way to do install is to create billing epoch:
https://github.com/cloudfoundry-incubator/cf-abacus/wiki/Install-Abacus#creating-a-billing-epoch-and-abacus-cf-bridges

I added a few things to consider in the Slack window wiki:
https://github.com/cloudfoundry-incubator/cf-abacus/wiki/Slack-Window

We use `SLACK=3d` since we want our invoices to be available on the 3rd day of the month.

**Comment by eujungkim** [comment]
@hsiliev Thank you for kind explanation.
Should To create billing epoch be done only first abacus installation time or every 1st day in month?

**Comment by hsiliev** [comment]
Billing epoch should be done only once - on the first install of Abacus

**Comment by eujungkim** [comment]
I'll try to change the SLACK value according to your guide.
Thank you and I'll close this issue.

**Comment by eujungkim** [comment]
I changed my SLACK value from 31D to 3D as your guide a few days ago and restart related applications.
But when I retrieve last row of abacus-aggregator-aggregated-usage collections for organization report today, some resource's day windows size is still 32. (Array[32])
Will the day window size be reduced to 3D automatically? Or should I do something to correct it?

I attached my mongodb document, and resource with resource_id linux-container has 32 day windows.

[report.txt](https://github.com/cloudfoundry-incubator/cf-abacus/files/1735958/report.txt)
